Fix placeholder contrast.

jasmussen opened the pull request:
This fixes #9502 and fixes #9532.

Screenshots:

![3-1 title 4 5-1 prompt](https://user-images.githubusercontent.com/1204802/44975314-7175c700-af62-11e8-8015-9a0ceaa30633.png)

Title is large text and needs 3:1 contrast. Placeholder is small text and needs 4.5:1.

![screen shot 2018-09-03 at 10 10 46](https://user-images.githubusercontent.com/1204802/44975353-7e92b600-af62-11e8-8ca8-7834f8248c16.png)

![screen shot 2018-09-03 at 10 12 09](https://user-images.githubusercontent.com/1204802/44975357-818da680-af62-11e8-876a-2cd7936a705c.png)

![screen shot 2018-09-03 at 10 14 17](https://user-images.githubusercontent.com/1204802/44975365-84889700-af62-11e8-9a3e-ad289af55350.png)

![screen shot 2018-09-03 at 10 14 35](https://user-images.githubusercontent.com/1204802/44975368-87838780-af62-11e8-9682-258185bedf4d.png)

jasmussen committed:
Fix placeholder contrast.

This fixes #9502 and fixes #9532.

youknowriad commented:
I'm well aware of the a11y guidelines but for me, it's too dark. There's no distinguishable difference between a placeholder and a regular text. (Try adding multiple empty paragraphs).

Differentiating between placeholders and text should also be an a11y guideline IMO.

kjellr commented:
> I'm well aware of the a11y guidelines but for me, it's too dark. There's no distinguishable difference between a placeholder and a regular text. (Try adding multiple empty paragraphs).

Fully agree here. The contrast between the text and placeholders in this PR is not significant enough to to differentiate between the two. 

If this is indeed the minimum Gray we want to use for placeholders, we should try pushing non-placeholder text to be darker instead. Changing it to 100% black helps just a little bit:

<img width="394" alt="screen shot 2018-09-03 at 11 06 19 am" src="https://user-images.githubusercontent.com/1202812/44981093-82760680-af69-11e8-9303-6ba3059e413e.png">

... but if I'm being honest, I'm not sure the difference there is noticeable enough either.

afercia commented:
> There's no distinguishable difference between a placeholder and a regular text

This is one of the many reasons why placeholders are so debated, and one more reason why they should be used sparingly. Unfortunately, design trends tend to use them as replacement for labels, [which is completely wrong](https://core.trac.wordpress.org/ticket/40331).

afercia commented on edit-post/assets/stylesheets/main.scss:
```diff
@@ -219,18 +219,18 @@ body.gutenberg-editor-page {
 		}
 		&::-moz-placeholder {
-			color: $dark-opacity-100;
+			color: $dark-opacity-300;
```
Worth noting Firefox by default applies also some opacity on the placeholder. In Gutenberg the computed opacity is `1` just because it's inherited by [the WP rule in forms.css](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/css/forms.css?rev=43309#L230). To make the Gutenberg styling really independent from core, there's the need to add `opacity: 1;` for any occurrence of the moz-placeholder.

afercia commented on edit-post/assets/stylesheets/main.scss:
```diff
@@ -219,18 +219,18 @@ body.gutenberg-editor-page {
-			color: $dark-opacity-100;
+			color: $dark-opacity-300;
 		}
 		&:-moz-placeholder {
```
Not sure why there's this version of the selector with one single colon. Shouldn't the one above with double colon be sufficient?

jasmussen replied:
Pretty sure that's copy pasta from https://stackoverflow.com/questions/19621306/css-placeholder-text-color-on-firefox where it's suggested that the single-colon version is necessary for Firefox 18 compatibility. Given we're on Firefox 61, and we have to support the last two versions, I think it's okay to drop that. Good catch.

afercia commented on packages/editor/src/components/post-title/style.scss:
```diff
@@ -32,9 +32,23 @@
+			color: $dark-opacity-200;
+		}
+		&::-moz-placeholder {
+			color: $dark-opacity-200;
```
see above about opacity

afercia commented on packages/editor/src/components/post-title/style.scss:
```diff
@@ -32,9 +32,23 @@
+		&:-ms-input-placeholder {
+			color: $dark-opacity-200;
+		}
+		&:-moz-placeholder {
```
see above, is this selector with single colon necessary?

afercia commented:
On a side note: now that there's the "Spotlight Mode", from an a11y perspective I wouldn't have objections if you really feel to lighten the placeholder color when Spotlight Mode is on. At one condition though: as said during the Spotlight Mode review, it shouldn't be on by default.

jasmussen commented:
Also fixes #8738

afercia commented:
Just noticed the File block placeholder would need a darker blue 🙂 

<img width="375" alt="screen shot 2018-09-07 at 16 50 15" src="https://user-images.githubusercontent.com/1682452/45226368-f0e6fb80-b2be-11e8-90f5-705a543e7a8f.png">

jasmussen commented:
Can we get a final review on this one (perhaps approval), so we can merge it in for 3.9?

jasmussen commented:
Rebased, squashed. 

Also pushed a fix to make the text darker:

![screen shot 2018-09-10 at 15 43 17](https://user-images.githubusercontent.com/1204802/45301453-4b779600-b511-11e8-8638-1baa33a214e0.png)

This is to ensure some contrast between basic text and placeholder text. It's still not ideal in situations like this:

![screen shot 2018-09-10 at 15 43 34](https://user-images.githubusercontent.com/1204802/45301477-57fbee80-b511-11e8-939c-76a4b8c05051.png)

But I feel like that can possibly be addressed separately. For example perhaps we could revisit the decision to show this placeholder text when the block isn't selected. Or put differently, maybe it's okay if an empty paragraph shows up as an empty paragraph until the caret is in it?

afercia commented:
Not sure if this fixes also captions displayed on top of images, see #5838 ?

jasmussen commented:
Not sure it does, but if it doesn't I'd like to look at that one separately.

kjellr commented:
Looking good. I think the darker text color helps. 👍 

> Not sure if this fixes also captions displayed on top of images, see #5838 ?

For all captions, this PR seems to eliminate the difference in color between placeholder and entered text:


_Image Block, placeholder:_
<img width="593" alt="screen shot 2018-09-10 at 10 35 42 am" src="https://user-images.githubusercontent.com/1202812/45304136-721fd780-b4e5-11e8-900a-fba719922e91.png">

_Image Block, caption:_
<img width="544" alt="screen shot 2018-09-10 at 10 35 53 am" src="https://user-images.githubusercontent.com/1202812/45304152-78ae4f00-b4e5-11e8-98df-9186990e0bd3.png">

_Gallery Block, placeholder:_
<img width="550" alt="screen shot 2018-09-10 at 10 36 11 am" src="https://user-images.githubusercontent.com/1202812/45304169-8237b700-b4e5-11e8-9f6c-c349c7620d0e.png">

_Gallery Block, caption:_
<img width="541" alt="screen shot 2018-09-10 at 10 36 22 am" src="https://user-images.githubusercontent.com/1202812/45304180-86fc6b00-b4e5-11e8-8ead-c738f1c0de85.png">

This does remove that accessibility issue, but I'm not sure if this was intended or not?

Commit:
Fix caption placeholder styles.

This fixes the caption styles for images (and embeds and anything else that has external captions), as well as for gallery items which has placeholder text overlaid. The overlay placeholder text has about a 4.8 average contrast ratio with the text on a scrim over a light image. This will vary a bit based on the darkness of the image.

jasmussen commented:
Okay so https://github.com/WordPress/gutenberg/pull/9562/commits/f04c24d0593eb0709566c727e2f717895a1a25bc makes the caption slightly darker. 

![caption color](https://user-images.githubusercontent.com/1204802/45348146-1f5e2280-b5ae-11e8-92ba-db343104ca6f.gif)

There's not much difference, but I think it's okay given that caption only appears when you've selected the image first.

jasmussen commented on packages/editor/src/components/rich-text/style.scss:
```diff
@@ -89,9 +89,18 @@
 	}
 
-	& + .editor-rich-text__tinymce {
-		opacity: 0.5;
```
@iseulde I'm realizing I may be causing regressions by introducing this rule. Can you give me a sanity check that removing this rule? 

What other cases might there be where this opacity would be used as the only way to indicate placeholder text?

iseulde replied:
I'm not sure. Only thing I can think of is adding emoji as placeholders, which would not look like placeholder text.

jasmussen commented:
Rebased, squashed. 

I could use a sanity check from @iseulde about the removal of this opacity rule: https://github.com/WordPress/gutenberg/pull/9562/files#diff-9bc45fdf28d434e26e152546f25fd08eL93 

Do you think removing it and relying on rgba color is going to have adverse effects? There was one effect as Kjell noted with the gallery images, but I fixed that in https://github.com/WordPress/gutenberg/pull/9562/commits/b9309eaaccc6d0c43ef5f3b540693e2bad50d81a

<img width="161" alt="screen shot 2018-09-11 at 10 41 17" src="https://user-images.githubusercontent.com/1204802/45348749-c1323f00-b5af-11e8-919f-1128bd349b44.png">

kjellr commented:
Looks great, thanks @jasmussen. 👏

aduth reviewed (approved):
Can't speak to accessibility or design, but code-wise you get my 👍
